hg-fast-export.py

```diff
@@ -127,52 +127,56 @@ def get_author(logmessage,committer,authors):
       return r
   return committer
 
+def remove_gitmodules(ctx):
+  """Removes all submodules of ctx parents"""
+  # Not quite sure whether submodules from all parents should be removed.
+  # Subrepos from first parent may be sufficient.
+  for parent_ctx in ctx.parents():
+    for submodule in parent_ctx.substate.keys():
+      wr('D %s' % submodule)
+  wr('D .gitmodules')
+
+def refresh_gitmodules(ctx):
+  """Updates list of ctx submodules according to .hgsubstate file"""
+  remove_gitmodules(ctx)
+  gitmodules=""
+  # Create the .gitmodules file and all submodules
+  for name,subrepo_info in ctx.substate.items():
+    gitRepoLocation=submodule_mappings[name] + "/.git"
+
+    # Populate the cache to map mercurial revision to git revision
+    if not name in subrepo_cache:
+      subrepo_cache[name]=(load_cache(gitRepoLocation+"/hg2git-mapping"),
+                           load_cache(gitRepoLocation+"/hg2git-marks",
+                                      lambda s: int(s)-1))
+
+    (mapping_cache,marks_cache)=subrepo_cache[name]
+    subrepo_hash=subrepo_info[1]
+    if subrepo_hash in mapping_cache:
+      revnum=mapping_cache[subrepo_hash]
+      gitSha=marks_cache[int(revnum)]
+      wr('M 160000 %s %s' % (gitSha,name))
+      sys.stderr.write("Adding/updating submodule %s, revision %s->%s\n"
+                       % (name,subrepo_hash,gitSha))
+      gitmodules+='[submodule "%s"]\n\tpath = %s\n\turl = %s\n' % (name,name,
+        submodule_mappings[name])
+    else:
+      sys.stderr.write("Warning: Could not find hg revision %s for %s in git %s\n" %
+        (subrepo_hash,name,gitRepoLocation))
+
+  if len(gitmodules):
+    wr('M 100644 inline .gitmodules')
+    wr('data %d' % (len(gitmodules)+1))
+    wr(gitmodules)
+
 def export_file_contents(ctx,manifest,files,hgtags,encoding='',plugins={}):
   count=0
   max=len(files)
-  for file in files:
-    if submodule_mappings and ctx.substate and file==".hgsubstate":
-      # Remove all submodules as we don't detect deleted submodules properly
-      # in any other way. We will add the ones not deleted back again below.
-      for module in submodule_mappings.keys():
-        wr('D %s' % module)
-
-      # Read .hgsubstate file in order to find the revision of each subrepo
-      data=ctx.filectx(file).data()
-      subHashes={}
-      for line in data.split('\n'):
-        if line.strip()=="":
-          continue
-        cols=line.split(' ')
-        subHashes[cols[1]]=cols[0]
```

Comment:
Removed parsing of `.hgsubstate` since `ctx.substate` already has all required info